### ISSUE TITLE
s3 bucket deployment time outs with many files

### ISSUE THREAD

**Issue by jeshan** [aws/aws-cdk#7571]
Deploying a website with `BucketDeployment` that has thousands of files times out as it needs to run longer than the lambda function timeout of 15 minutes.

### Reproduction Steps
1. Create a website with thousands of files
2. Deploy as follows:
```javascript
new BucketDeployment(this, "WebsiteDeployment", {
        sources: [Source.asset(placeHolderSource)],
        destinationBucket: websiteBucket,
        distribution,
        retainOnDelete: false
      });
```

#### The serverless function
```
REPORT RequestId: 3e301853-5a0a-4220-a93f-eb9b03bdcfa4 Duration: 900004.88 ms Billed Duration: 900000 ms Memory Size: 128 MB Max Memory Used: 128 MB
```

#### the cloudformation stack
![image](https://user-images.githubusercontent.com/1020565/80136457-c5f40d00-85b2-11ea-9a53-47a1bfa5d16c.png)

### Environment

  - **CLI Version      :** 1.34.1
  - **Framework Version:**
  - **OS               :** Linux
  - **Language         :** JavaScript

### Other

# Suggestions
I think it's because of the s3 sync is checking thousands of times if the files exist.

https://github.com/aws/aws-cdk/blob/b30add8c70dcf6239c137a38eb764513ce1e1f29/packages/%40aws-cdk/aws-s3-deployment/lambda/src/index.py#L133-L134

1. I'm thinking there could be an alternative or an option for cdk users to use the `cp` command instead, or:
2. the team can make it use `aws s3 cp --recursive` when creating the deployment and continue using `sync` for subsequent updates. That way, at least the cdk stack will stabilise (but postpones dealing with the issue)

--- 

This is :bug: Bug Report

**Comment by jeshan** [comment]
I saw that we can configure memory on the deployment resource to increase the lambda performance. Is that the way to go?

**Comment by armandosoriano** [comment]
I'm facing the same, it logs copying file until this line
`Completed 109.5 MiB/261.8 MiB (4.7 MiB/s) with 1 file(s) remaining `
then 
```
[ERROR]	2020-05-04T00:35:03.737Z	529deb7e-e023-4b70-8d42-82a77cc0611e	Command '['python3', '/var/task/aws', 's3', 'cp', 's3://cdktoolkit-xxxxxxxxx/assets/x.zip', '/tmp/tmpwl1hisdf/dca9c054-1839-42fa-a5f1-1d702f2e1167']' died with <Signals.SIGKILL: 9>.
```

**Comment by armandosoriano** [comment]
@jeshan I just tried increasing memory to `256`mb  and it thrown `[Errno 28] No space left on device`, looks like Python is struggling

**Comment by iliapolo** [comment]
@jeshan Wrote: 

> I think it's because of the s3 sync is checking thousands of times if the files exist.

I am not versed enough in how the CLI optimizations work, but its possible. In any case, we already have a [feature request](https://github.com/aws/aws-cdk/issues/953) for toggling the `--delete` flag, which I assume is what might be causing the behavior you are describing. 

Feel free to 👍 it!

@armandosoriano Wrote:

> @jeshan I just tried increasing memory to `256`mb and it thrown `[Errno 28] No space left on device`, looks like Python is struggling

Dont think this its related to python per se. 
I suspect that increasing the memory just made the download work a little faster, and thus giving it enough time to hit the [500MB Limit](https://github.com/aws/aws-cdk/issues/6893). Unfortunately that limitation is harder to workaround. 
So if your deployment is larger than 500MB you'll need to manually run `aws cp` or `aws sync` from your CI machine for now, until we come up with a solution for this.

In any case, the 15 minutes time limit is a hard [lambda limit](https://docs.aws.amazon.com/lambda/latest/dg/gettingstarted-limits.html), so no matter how many optimizations or configurations we provide, if the `sync/cp` command takes longer than that, its going to be a problem.

I am leaning towards closing this issue as it seems to relate to already existing issues.

Thoughts?

**Comment by jeshan** [comment]
before closing, let's share ideas that's beyond the scope of #953

Instead of using `aws s3 cp`, we could do what it's doing in parallel from the python code (using regular python multi-threading). I'm not sure how much this is of a good idea.

> so no matter how many optimizations or configurations we provide

Yes, but at least we'll be able to make progress in the right direction by quite a bit.

**Comment by armandosoriano** [comment]
@iliapolo our deployment weights less than 300MB so I'm not sure if it is really related to that limit.

Indeed I'm just guessing but it looks like python process gets stuck during the copy files operation, while it still reports 5-8MB/s speed, which is indeed faulty. 

Maybe the copy is proper but for some reason the process is not ending.
Maybe the process ends but it is not reported and thus the timeout is reached.

Without more knowledge about how that is being managed is difficult to know. Hope those guessings can at least help you in your investigation.

**Comment by iliapolo** [comment]
@armandosoriano @jeshan thanks for the feedback.

Keeping this on our docket as we discuss what the best course of action here would be.

**Comment by iliapolo** [comment]
Closing in favor of https://github.com/aws/aws-cdk/issues/7950
